### DISCUSSION TITLE
PURL URL is not identified in Dependency Tracker

### DISCUSSION THREAD

**Discussion by VASAVI512** [DependencyTrack/dependency-track#1829] [General]
Hello Steve,

Dependency tracker is not identifying the PULR url, though they correct. But when we i have given the CPE manually in DT its showing the component Vulnerabilities. SBOM json file has PULR information.

Example PURL urls :
"purl" : "pkg:deb/debian/linux-libc-dev@4.19.194-2?arch=amd64",
"purl" : "pkg:deb/debian/libsasl2-modules-otp@2.1.27%2Bdfsg-1%2Bdeb10u1?arch=i386",
"purl" : "pkg:deb/ubuntu/libdb5.3-dev@5.3.28%2Bdfsg1-0.5?arch=amd64",

DT version: 4.5.0
SBOM generated from Blackduck Cyclone DX 1.3 version

Please help us to understand, why DT is not showing few of the PURL vulnerabilities?.

Regards,
Vasavi

**Comment by JN-CSIRT** [comment]
The route to DEB repository (as well as GitHub Repository) does not exist in DT. There are selected repositories only. 
It should be added into DT.

**Comment by VASAVI512** [comment]
Thank you for your response. 
Where can i get the list of repositories that DT will add through PURL link. 
Is there any plan to add/update the repositories ?

**Comment by nscuro** [comment]
Dependency-Track per default uses Sonatype OSS Index to identify vulnerabilities based on PURL. OSS Index however does not support the Debian ecosystem (see [here](https://ossindex.sonatype.org/ecosystems) for supported ecosystems).

When you provide CPEs, Dependency-Track can match those against the [NVD](https://nvd.nist.gov/), which *does* include Debian vulnerabilities. The NVD doesn't support PURL, that's why components with PURL cannot be matched directly against the NVD right now.

**Reply by g-sahil22** [reply]
Hi @nscuro ,

I generate SBOM using cdxgen for live OS based on RPM, and sonatype OSS index support RPM, but still dependency track no report any vulnerability 

pkg:rpm/NetworkManager@1.40.0#NetworkManager-1.40.0-5.0.2.el8_7.src.rpm
pkg:rpm/abattis-cantarell-fonts@0.0.25#abattis-cantarell-fonts-0.0.25-6.el8.src.rpm
pkg:rpm/bash@4.4.20#bash-4.4.20-4.el8_6.src.rpm
pkg:rpm/binutils@2.30#binutils-2.30-117.0.3.el8.src.rpm

Thanks
Sahil

**Reply by syalioune** [reply]
Hello,

You can refer to nscuro reply here https://github.com/DependencyTrack/dependency-track/discussions/1829#discussioncomment-3229510

OSS Index does not return any vulnerabilities for those packages, so does DT.

![image](https://user-images.githubusercontent.com/6144741/222433909-9bcae170-7efa-4157-8bf2-e7acc1eb2cf7.png)

**Comment by VASAVI512** [comment]
Understood. Yes if i provide the CPEs manually DT is able to show the vulnerabilities for deb components as well.

But, the below PURL urls also not working in DT, though they are Sonatype OSS index range. Conda is supported by Sonatype. 
Please let me know understand why these are PURL urls are not showing risk in DT. 

Python Pillow, 8.3.0
"purl" : "pkg:conda/pillow@8.3.0-py37h0f21c89_0-linux-64"

Python programming language , 2.7.16
"purl" : "pkg:rpm/clearlinux/python-core@2.7.16-123?arch=x86_64"

Flask-Cors, 3.0,10
"purl" : "pkg:pypi/flask-cors@3.0.10"

python-sqlite3, 2.7.16
"purl" : "pkg:yocto/python-sqlite3@2.7.16-r0?arch=core2-32",

Jinja, 2.1
"purl" : "pkg:apk/alpine/py3-jinja2@2.10-r3?arch=noarch",

**Reply by nscuro** [reply]
Well, OSS Index will not always have all possible vulnerabilities. If OSS Index doesn't report vulnerabilities, Dependency-Track can't show them.

You can test your PURLs against OSS Index on their website: https://ossindex.sonatype.org/search

Here's an example of a PURL that works: https://ossindex.sonatype.org/component/pkg:maven/com.fasterxml.jackson.core/jackson-databind@2.13.2

Here's the result for `Flask-Cors`: https://ossindex.sonatype.org/component/pkg:pypi/flask-cors@3.0.10

You can report missing or wrong vulnerabilities to Sonatype, see here: https://ossindex.sonatype.org/doc/report-vulnerability

**Reply by g-sahil22** [reply]
HI @nscuro ,

Which parameter dependency track is used in backed for google OSV?

**Reply by nscuro** [reply]
@g-sahil22 Can you expand on that question? I'm not sure I understand what you're asking for.

**Reply by g-sahil22** [reply]
for NVD it is CPE for the Sonatype OSS index it is PURL, similarly, for Google OSV, because google OSV contains a good database of OS packages

**Reply by nscuro** [reply]
Got it, thanks. For OSV, DT will also use PURL.

**Reply by g-sahil22** [reply]
Thanks @nscuro

**Comment by JN-CSIRT** [comment]
Hi Niklas,
what's about Github and Sourceforge PURLs? It could be great to check the components in Github for actuality as e.g. for Maven...
OSS does not support them in Ecosystem :( 
Is there any plans to add them into DT?

**Reply by nscuro** [reply]
No active plan for it, but if you raise an enhancement request we can track it and deliver that in a future version.

**Comment by VASAVI512** [comment]
Hi Niklas,

I am raising the ticket wit Sonatype, though there are CPE's we do not find the vulnerabilities in DT.
 And,
Comp: Git
Version: 2.20.1
pkg:deb/debian/git@2.20.1-2%2Bdeb10u3?arch=amd64&epoch=1
We see the vulnerablility data in DT for the DEB packages, how is it possible, as per your comments deb is not in supported ecosystems.

Regards,
Vasavi

**Reply by nscuro** [reply]
Debian *was* supported by OSS Index prior to their big migration (see the announcement [here](https://ossindex.sonatype.org/updates-notice)):

> Some ecosystems will no longer be supported.As part of this change we will be dropping support for Drupal, **Debian**, Chocolatey, Alpine, Bower and Go Dep. Note: Go Mod will continue to be supported and we encourage all OSS Index users to upgrade to newer Go Mod modules.

 It's possible that the vulnerability you're referring to was identified prior to this change.

**Comment by officerNordberg** [comment]
I worked on the fuzzy matching precisely to match Debian artifacts and my disappointment with PURL matching of NVD CVEs. YMMV

**Comment by VASAVI512** [comment]
Hi Niklas,

We would like to know how DT data back up and restore is working. Can you share any document how to do Back up and restore through Dockers.

Regards,
Vasavi

**Comment by Gowtham-Rangasamy** [comment]
Hi Everyone,

Recently we have deployed Dependency track in our Organization. DT is unable to identify the Vunerabilities in the components.
Pushed the components in purl format as CycloneDX XML format
<property name="purl">pkg:nuget/jQuery.Validation@1.17.0</property>
<property name="purl">pkg:nuget/Microsoft.Owin.Security.OpenIdConnect@4.0.1</property>
<property name="purl">pkg:nuget/Microsoft.Owin.Security@4.0.1</property>
none of the components in the DT across the organization is identified with the vulnerabilities.
is there any thing which I am missing here.

Regards,
Gowtham R
